query-parser/src/util.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+extern crate edn;
+extern crate mentat_query;
+
+use std::collections::BTreeMap;
+
+use self::edn::Value::PlainSymbol;
+use self::mentat_query::Variable;
+
+use super::error::{FindParseError};
+
+/// If the provided EDN value is a PlainSymbol beginning with '?', return
+/// it wrapped in a Variable. If not, return None.
+pub fn value_to_variable(v: &edn::Value) -> Option<Variable> {
+    if let PlainSymbol(ref sym) = *v {
+        if sym.0.starts_with('?') {
+            return Some(Variable(sym.clone()));
+        }
+    }
+    return None;
+}
+
+/// If the provided slice of EDN values are all variables as
+/// defined by `value_to_variable`, return a Vec of Variables.
+/// Otherwise, return an error.
+pub fn values_to_variables(vals: &[edn::Value]) -> Result<Vec<Variable>, FindParseError> {
+    let mut out: Vec<Variable> = Vec::with_capacity(vals.len());
+    for v in vals {
+        if let Some(var) = value_to_variable(v) {
+            out.push(var);
+            continue;
+        }
+        return Err(FindParseError::InvalidInput(v.clone()));
+    }
+    return Ok(out);
+}
+
+#[test]
+fn test_values_to_variables() {
+    // TODO
+}
+/// Take a slice of EDN values, as would be extracted from an
+/// `edn::Value::Vector`, and turn it into a map.
+///
+/// The slice must consist of subsequences of an initial plain
+/// keyword, followed by one or more non-plain-keyword values.
+///
+/// The plain keywords are used as keys into the resulting map.
+/// The values are accumulated into vectors.
+///
+/// Invalid input causes this function to return `None`.
+pub fn vec_to_keyword_map(vec: &[edn::Value]) -> Option<BTreeMap<edn::Keyword, Vec<edn::Value>>> {
+    let mut m = BTreeMap::new();
+
+    if vec.is_empty() {
+        return Some(m);
+    }
+
+    if vec.len() == 1 {
+        return None;
+    }
+
+    // Turn something like
+    //
+    //   `[:foo 1 2 3 :bar 4 5 6]`
+    //
+    // into
+    //
+    //   `Some((:foo, [1 2 3]))`
+    fn step(slice: &[edn::Value]) -> Option<(edn::Keyword, Vec<edn::Value>)> {
+        // This can't be right -- we can't handle [:foo 1 2 3 :bar].
+        if slice.len() < 2 {
+            return None;
+        }
+
+        // The first item must be a keyword.
+        if let edn::Value::Keyword(ref k) = slice[0] {
```

Comment:
I did; the split functions produce iterators, which makes the code a little more confusing. We've already validated the length above.

(What I really want is slice destructuring…)